pallets/marketplace/src/lib.rs

```diff
@@ -768,7 +771,49 @@ pub mod pallet {
                     <StoredJobStatus<T>>::remove(&job_id.0, &job_id.1);
                     <StoredJobRegistration<T>>::remove(&job_id.0, &job_id.1);
                 }
-                _ => Err(Error::<T>::JobRegistrationUnmodifiable)?,
+                JobStatus::Matched => {
+                    T::MarketplaceHooks::finalize_job(job_id, T::RewardManager::refund(job_id))?;
+
+                    // Remove matching data
+
+                    for (p, _) in <AssignedProcessors<T>>::iter_prefix(&job_id) {
+                        <StoredMatches<T>>::remove(&p, &job_id);
+                        <StoredStorageCapacity<T>>::remove(&p);
+                    }
+
+                    let _ =
+                        <AssignedProcessors<T>>::clear_prefix(&job_id, T::MaxSlots::get(), None);
+                    <StoredJobStatus<T>>::remove(&job_id.0, &job_id.1);
+                    <StoredJobRegistration<T>>::remove(&job_id.0, &job_id.1);
+                }
+                JobStatus::Assigned(assigned) => {
+                    let remaining_reward = Self::reserved(job_id);
+                    let reward_per_processor = remaining_reward
+                        .checked_div(&(assigned.into()))
```

Comment:
if you intentionally distribute between the already assigned processors (so more than they would receive by fulfilling), maybe add a comment explaining that.
I expected the non-assigned ones to be payed back to consumer

Reply:
This was not intentional, fixed in the latest commit.